File: leetcode/python/0105. Construct Binary Tree from Preorder and Inorder Traversal.py
# Time complexity: O(N)
# Space complexity: O(H), where H is the height of the tree

class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution(object):
    def buildTree(self, preorder, inorder):
        if not preorder or not inorder:
            return None
        
        head = preorder[0]
        mid = inorder.index(head)
        res = TreeNode(val=head)

        res.left = self.buildTree(preorder[1:mid+1], inorder[:mid])
        res.right = self.buildTree(preorder[mid+1:], inorder[mid+1:])

        return res

File: leetcode/java/0105. Construct Binary Tree from Preorder and Inorder Traversal.java
package leetcode.java;

import java.util.Arrays;
import java.util.stream.IntStream;

// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

// Time complexity: O(N)
// Space complexity: O(H), where H is the height of the tree

class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if (preorder.length > 0 && inorder.length > 0) {
            TreeNode res = new TreeNode(preorder[0]);
            int pivot = this.indexOf(inorder, preorder[0]);
            res.left = this.buildTree(Arrays.copyOfRange(preorder, 1, pivot), Arrays.copyOfRange(inorder, 0, pivot));
            res.right = this.buildTree(Arrays.copyOfRange(preorder, pivot+1, preorder.length), Arrays.copyOfRange(inorder, pivot+1, inorder.length));
            return res;
        }
        return null;
    }

    public int indexOf(int[] arr, int target) {
        return IntStream.range(0, arr.length)
            .filter(x -> target == arr[x])
            .findFirst()
            .orElse(-1);
    }
}
